File: java/pjq/leecode/AddTwoNumbers.java
package pjq.leecode;

/**
 * 两数相加：
 */
public class AddTwoNumbers {

    public static void main(String[] args) {
        AddTwoNumbers addTwoNumbers = new AddTwoNumbers();
        int arr[] = {1, 2, 3};
        int arr1[] = {8, 8, 8};
        ListNode l1 = addTwoNumbers.addNode(arr);
        ListNode l2 = addTwoNumbers.addNode(arr1);

        ListNode res = addTwoNumbers.addTwoNumbers(l1, l2);
        while (res != null) {
            System.out.print(res.val + "->");
            res = res.next;
        }

    }

    public ListNode addNode(int[] arr) {
        ListNode listNode = new ListNode();
        listNode.val = arr[0];
        ListNode nextNode = listNode;
        for (int i = 1; i < arr.length; i++) {
            ListNode tmpNode = new ListNode();
            tmpNode.val = arr[i];
            nextNode.next = tmpNode;
            nextNode = tmpNode;
        }
        return listNode;
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode res = new ListNode(-1);
        ListNode node = res;
        int addFlag = 0;
        int sum = 0; //进位
        while (l1 != null || l2 != null) {
            ListNode tmp = null;
            int num1 = (l1 == null) ? 0 : l1.val;
            int num2 = (l2 == null) ? 0 : l2.val;
            sum = num1 + num2 + addFlag;

            addFlag = sum >= 10 ? 1 : 0;//更新进位

            tmp = new ListNode(sum % 10);
            node.next = tmp;
            node = tmp;

            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }

        if (sum >= 10) {
            node.next = new ListNode(1);
        }

        return res.next;
    }

}

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }
}

File: php/pjq.leetcode/AddTwoNumbers.php
<?php

/**
 * 两数相加
 * @desc https://leetcode-cn.com/problems/add-two-numbers/
 * Class AddTwoNumbers
 */

class AddTwoNumbers
{

    function __construct()
    {
    }

    function addTwoNumbers($l1, $l2)
    {
        $root = new ListNode(-1);
        $node = $root;
        $i = 0;  //进位
        $sum = 0;

        while ($l1 != null || $l2 != null) {
            $tmp = null;
            $num1 = ($l1 == null) ? 0 : $l1->val;
            $num2 = ($l2 == null) ? 0 : $l2->val;
            $sum = $num1 + $num2 + $i;

            $i = $sum >= 10 ? 1 : 0;


            $tmp = new ListNode($sum % 10);

            $node->next = $tmp;
            $node = $node->next;

            if ($l1 != null) $l1 = $l1->next;
            if ($l2 != null) $l2 = $l2->next;
        }

        if ($sum >= 10) {
            $node->next = new ListNode(1);
        }
        return $root->next;
    }

    public function createListNode1() {
        $node1 = new ListNode(2);
        $node2 = new ListNode(4);
        $node3 = new ListNode(3);

        $node1->next = $node2;  $node2->next = $node3;
        return $node2;
    }
    public function createListNode2() {
        $node1 = new ListNode(5);
        $node2 = new ListNode(6);
        $node3 = new ListNode(4);

        $node1->next = $node2;  $node2->next = $node3;
        return $node2;
    }
}

class ListNode
{
    public $val = 0;
    public $next = null;

    function __construct($val = 0, $next = null)
    {
        $this->val = $val;
        $this->next = $next;
    }
}

$obj = new AddTwoNumbers();
$node1 = $obj->createListNode1();
$node2 = $obj->createListNode2();
$res_node = $obj->addTwoNumbers($node1,$node2);
while($res_node != null) {
    echo $res_node->val . " ";
    $res_node = $res_node->next;
}

